tools/build/install-core-dist.raku

```diff
@@ -1,6 +1,7 @@
+BEGIN %*ENV<RAKUDO_INSTALL_STAGING> = 1;
```

Comment:
We cannot depend on an env var to choose a repo like this — repos are chosen via their repo spec and adding it to the repository chain

Reply:
This is not about choosing a repo but about working around detecting the current distribution at compile time. Since this is a kind of bootstrapping process an exception should not be a problem for it.

Reply:
I would try to see if there is another solution, but an explicit flag for staging makes things a lot simpler and easier to understand.

Reply:
Addressed with 8006fdab0808c5ec7e73ca4a9c99b668d3c8a5e8

Reply:
What about debian packagers that use this same CURS workflow for creating os packages?

Reply:
I need to see what they do. Better if there is a test for it.

Reply:
@ugexe I see no solution for this. The problem is about conflicting understanding of where `core` repo is located between the script and precompiling `rakudo` sub-process.

For the script it is the staging repo. But when it launches precompilation the repository is reported with its spec as `CompUnit::Repository::Staging#name(core)#<path>` in `RAKUDO_PRECOMP_WITH`. That resolves into `CompUnit::Repository::Unknown` since staging module is not available.

Consequently, the precompilation resolves core repository with the default prefix for staging tools (Rakudo home) which is `$BASE_DIR/gen/build_rakudo_home`. This one remains empty. I'm not sure what is the purpose of it, but @patrickbkr may know better.

The result of the conflict is the compiler failing to locate the actual meta file for the core libraries because the script installs it into the staging repository. This was never a problem because the distribution was not used. But this PR tries to install `$?DISTRIBUTION` as early as possible and reveals the conflict.

Possible solutions would be:

- make the staging repo report itself as `inst#name(core)` for precompilation
- provide a way for `CU::RepositoryRegistry.repository-for-spec` to load custom repository classes by full module file path

Neither approach doesn't appeal to me as I anticipate unforeseen side effects. Though the first one feels as the most reliable because precompilation must not care about staging or not staging, all it needs is access to other precompiled modules, perhaps.

And yet, considering that the core libraries do not need the distribution, I would stick to the controlling environment variable. Just rename it for self-documenting purposes. Say, call it `RAKUDO_NO_PRECOMP_DIST`. The meaning would remain the same: when on, no attempt to locate the distribution would be done at compile time, except for the apparent locations in CU code where it is known as a result of resolution algorithms.

With regard to distro builders, they must not be affected unless they use the same mocking repository, as Rakudo build does.

Reply:
Eventually, after playing on with the problem, I came to the conclusion that explicit blocking of distribution deduction is currently the only solution. Unfortunately, since the precomp parent process provides distribution information to its children via `core` repository, which is then resolved to `$(BASE_DIR)/gen/build_rakudo_home` by them, whereas the required meta is stored under the real installation path, there is no non-hacky solution to get the task done.

Yet, the modules we install via staging process do not need the distribution information. Would they need it we'd hit the problem much earlier. Therefore disabling the deduction would be totally legit.

To be noted that, in my view, any solution to the conflict would be a temporary workaround for a hack, considering [the comment](https://github.com/rakudo/rakudo/blob/90709d5334d59f2e292fd2ec765f1b9a45695b6c/tools/templates/moar/rakudo-m-build.c.in#L40-L43).

Eventually, I ended up renaming the environment variable from rather obscure `RAKUDO_INSTALL_STAGING` to `RAKUDO_NO_PRECOMP_DIST`, as suggested in the previous comment. I consider this the final solution unless the whole staging process finds a non-hacky approach.

@ugexe would you still mind merging?

Reply:
I have two concerns with merging this in the immediate future.

1) There are two blockers that suggest precompilation was broken/deoptimized during the last round of `CompUnit::*` refactorings. As it stands we can't know if this PR would cause similar issues, because the PR is based on branch containing the issue itself. It would probably be wise to wait for that to be fixed / the next release.

2) I sort of wonder if this idea in general needs further design considerations. For instance: how is this supposed to interact with a theoretical `emulates` or `supercedes` etc features (see: https://github.com/Raku/old-design-docs/blob/master/S22-package-format.pod). Similarly, at many of my jobs our codebases would contain something like `lib/DBIx/Class/SomeDBICModule.pm` which was a patched version of the file that would be used by the rest of the original `DBIx-Class` distribution when it does `use DBIx::Class::SomeDBICModule`. My concern is then how is this behavior -- overriding a single module in another distribution -- supposed to be done in a future version of Raku? My mental model has been that `auth` would be the way to handle that (see https://stackoverflow.com/questions/55671684/how-does-raku-decide-which-version-of-a-module-gets-loaded#comment130257025_55671685), i.e. I wouldn't have pinned `auth` if I was the one implementing this. What are your thoughts on this?

Reply:
> 2\. For instance: how is this supposed to interact with a theoretical `emulates` or `supercedes` etc features

I don't see a reason why should it somehow prevent these two from being feasible. Though I may overlook something. But, eventually, It's like having an extra `provides`, just a somewhat more complex one.

> 2\. My mental model has been that `auth` would be the way to handle that

It would be nice to have a test for that. Because yes, I see it the same: without `auth` it would chose based upon `ver` and `api` by both matching and sorting for higher versions. And if it wouldn't then I'd consider it a bug.

> 1. because the PR is based on branch containing the issue itself.

This is the primary reason to postpone merging and, perhaps, any further work on this subject. Knowing the usual course of things, I predict someone would have to come up with an alternative implementation later when things are  settled down.

BTW, there was a note on packaging how that would be affected with staging issues. I was thinking about it and found no reason for them to fail. Would the process be susceptible to the problem then modules using `$?DISTRIBUTION` or `%?RESOURCES` would already be affected.

Reply:
> I don't see a reason why should it somehow prevent these two from being feasible. Though I may overlook something. But, eventually, It's like having an extra provides, just a somewhat more complex one.

There is a closed PR by me implementing emulates, and indeed its fairly simple. However, if distribution `Foo::Dist` contains a `JSON::Fast` and some other dist contains a META6.json with the following, then it seems like `Foo::Dist` maybe (I honestly don't know) should load the emulated `JSON::Fast` and not its own.

```
# Not::Foo::Dist
"emulates" : {
  "JSON::Fast" : "JSON::XS:auth<cpan:MLEHMANN>"
}
```

---

> Because yes, I see it the same: without auth it would chose based upon ver and api by both matching and sorting for higher versions. 

Actually my entire point here was a bit off. I don't think `auth` is useful for solving the issue I mentioned... rather something like: the ability to put a single CURFS `lib/JSON/Fast.rakumod` at the head of the repo chain such that any downstream distributions that try to load a `JSON::Fast` (even if they include their own) would load the CURFS one mentioned earlier. I'm not sure this type of overriding could ever work, though, due to issues like inferring if some file should get its own `$?DISTRIBUTION` or that of whatever its overriding.

Reply:
> then it seems like `Foo::Dist` maybe (I honestly don't know) should load the emulated `JSON::Fast` and not its own.

> the ability to put a single CURFS `lib/JSON/Fast.rakumod` at the head of the repo chain such that any downstream distributions that try to load a `JSON::Fast` (even if they include their own) would load the CURFS one mentioned earlier.

Actually, that's what I thought about. Moreover, having a CURFS at the head of the list is what logically comes to mind first. So far, the approach of using `$*REPO.head.need($dep-spec)` should implemented this logic (see CURR.NEED in this PR).

BTW, I think that if a distro has a module then any other module of the distro must load it's local version in first place. You know, a principle of lexical scoping applied to distributions. It's currently broken, but I would consider this rule as a must have. Including the case of `Foo::Dist` loading its own implementation of `JSON::Fast`.

Reply:
An unrelated consideration: we can have a caching repo at the top of the chain. Provided dependency specification has a unique ID, it can speed up lookups.